### PR TITLE
feat: add recentlyViewed 

### LINKED ISSUE #1340: Support quick search recently opened

### PR DESCRIPTION
**Pull request by JimmFly**
Close #1340

### CODE REVIEW

File: apps/web/src/hooks/affine/use-recent-views.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  const [workspace] = useCurrentWorkspace();
+  const workspaceId = workspace?.id || '';
+  const blocksuiteWorkspace = workspace?.blockSuiteWorkspace || null;
+  const id = router.query.pageId as string;
+  const meta = usePageMeta(blocksuiteWorkspace).find(meta => meta.id === id);
+  const [recentlyViewed, setRecentlyViewed] = useAtom(workspaceRecentViewsAtom);
+
+  useEffect(() => {
+    router.events.on('routeChangeComplete', () => {
+      if (router.query.pageId) {
```
Comment (lawvs) [inline apps/web/src/hooks/affine/use-recent-views.ts:20]:
```suggestion
      if (id) {
```

File: apps/web/src/hooks/affine/use-recent-views.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import { WorkspaceRecentViews, workspaceRecentViewsAtom } from '../../atoms';
+import { useCurrentWorkspace } from '../current/use-current-workspace';
+import { usePageMeta } from '../use-page-meta';
+
+export function useRecentlyViewed() {
+  const router = useRouter();
+  const [workspace] = useCurrentWorkspace();
+  const workspaceId = workspace?.id || '';
+  const blocksuiteWorkspace = workspace?.blockSuiteWorkspace || null;
+  const id = router.query.pageId as string;
```
Comment (lawvs) [inline apps/web/src/hooks/affine/use-recent-views.ts:14]:
```suggestion
  const id = router.query.pageId;
```

File: apps/web/src/hooks/affine/use-recent-views.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        ];
+
+        setRecentlyViewed({
+          ...recentlyViewed,
+          [workspaceId]: newRecentlyViewed,
+        });
+      }
+    });
+    return () => {
+      router.events.off('routeChangeComplete', () => {});
```
Comment (lawvs) [inline apps/web/src/hooks/affine/use-recent-views.ts:36]: Please dispose listener corrent

File: apps/web/src/components/pure/workspace-slider-bar/index.tsx [hunk 4]
```diff
@@ -118,6 +119,7 @@ export const WorkSpaceSliderBar: React.FC<WorkSpaceSliderBarProps> = ({
       openPage(pageId);
     }
   }, [createPage, openPage]);
+  useRecentlyViewed();
```
Comment (himself65) [inline apps/web/src/components/pure/workspace-slider-bar/index.tsx]: does this necessary?

File: apps/web/src/hooks/affine/use-recent-views.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        {
+          title: meta.title || 'Untitled',
+          id: pageId as string,
+          mode: meta.mode || 'page',
+        },
+        ...workspaceRecentlyViewed
+          .filter(item => item.id !== pageId)
+          .slice(0, 2),
+      ];
+      console.log('newRecentlyViewed', newRecentlyViewed);
```
Comment (himself65) [inline apps/web/src/hooks/affine/use-recent-views.ts]: remove

File: apps/web/src/atoms/index.ts [hunk 6]
```diff
@@ -60,3 +59,13 @@ export const workspacesAtom = atom<Promise<RemWorkspace[]>>(async get => {
   );
   return workspaces.filter(workspace => workspace !== null) as RemWorkspace[];
 });
+
+export type WorkspaceRecentViews = Record<
+  string,
+  { title: string; id: string; mode: 'page' | 'edgeless' }[]
+>;
+
+export const workspaceRecentViewsAtom = atomWithStorage<WorkspaceRecentViews>(
```
Comment (himself65) [inline apps/web/src/atoms/index.ts:67]: modify the atom here to support only 3 items

**Review by Himself65 — CHANGES_REQUESTED** [review]

**Review by Himself65 — APPROVED** [review]

**Pull request merged**